src/main/java/au/org/emii/ncdfgenerator/NcdfEncoderBuilder.java

```diff
@@ -0,0 +1,1476 @@
+			}
+
+			if( dimensions == null )
+			{
+				dimensions = new HashMap< String, IDimension> ();
+			}
+
+			if( name != null
+				&& encodeValue != null
+				&& attributes != null )
+			{
+				System.out.println( "whoot creating encoder " + name  );
+
+				return new VariableEncoder ( name , new ArrayList<IDimension>(dimensions.values()), encodeValue , attributes ) ;
+			}
+			else {
+				throw new NcdfGeneratorException( "missing something" );
+				// return null;
+			}
+		}
+		return null;
+	}
+
+
+	private Map< String, IVariableEncoder> parseVariableEncoders( Node node, Map< String, IDimension> dimensionsContext  )
+		throws NcdfGeneratorException
+	{
+		if( isNodeName( node, "variables"))
+		{
+			Map< String, IVariableEncoder> m = new HashMap < String, IVariableEncoder>();
+			for( Node child : new NodeWrapper(node)) {
+				IVariableEncoder e = parseVariableEncoder( child, dimensionsContext  );
+				if( e != null )
+					m.put( e.getName(), e );
+			}
+			return m;
+		}
+		return null;
+	}
+
+	private Map< String, String> parseSource( Node node )
+	{
+		if( isNodeName( node, "source")) /// data or dataSource
+		{
+			Map< String, String> source = parseKeyVals( node);
+
+			/*
+			System.out.println( "*** GOT CONFIG " +  source.size()   );
+			for( Map.Entry< String, String> entry : source.entrySet()) {
+				System.out.println( "*** " + entry.getKey() + " " + entry.getValue() );
+
+			} */
+			return source;
+		}
+		return null;
+
+	}
+
+	NcdfDefinition parseDefinition( Node node )
+		throws NcdfGeneratorException
+	{
+		// think we need a context?
+		if( isNodeName( node, "definition"))
+		{
+			Map< String, String> source = null;
+			Map< String, IDimension> dimensions = null;
+			Map< String, IVariableEncoder> encoders = null;
+
+			// pick out dimensions
+			for( Node child : new NodeWrapper(node)) {
+
+				if( source == null )
+					source = parseSource( child );
+
+				if( dimensions == null )
+					dimensions = parseDimensions( child );
+
+				if( encoders == null )
+					encoders = parseVariableEncoders( child, dimensions );
+			}
+
+			String schema = source.get( "schema" );
+			String virtualDataTable = source.get( "virtualDataTable" );
+			String virtualInstanceTable =source.get( "virtualInstanceTable" );
+
+			return new NcdfDefinition( schema, virtualDataTable, virtualInstanceTable, dimensions, encoders );
+		}
+		return null;
+	}
+}
+
+
+class NcdfEncoder
+{
+	final IExprParser exprParser;
+	final IDialectTranslate translate ;
+	final Connection conn;
+	final ICreateWritable createWritable; // generate a writable
+	final NcdfDefinition definition ;
+	final String filterExpr;
+
+	final int fetchSize;
+	IExpression selection_expr;
+	ResultSet featureInstancesRS;
+
+	public NcdfEncoder(
+		IExprParser exprParser,
+		IDialectTranslate translate,
+		Connection conn,
+		ICreateWritable createWritable,
+		NcdfDefinition definition,
+		String filterExpr
+	) {
+		this.exprParser = exprParser;
```

Comment:
I initially thought exprParser was for passing the subset parameters.  Perhaps renaming it would make the code clearer e.g to NetcdfDefinitionParser or NetcdfTemplateParser